### PR TITLE
Add included_domains and included_entities

### LINKED ISSUE #49: Add include_domains and include_entities
**Is your feature request related to a problem? Please describe.**
Currently, the configuration file allows for excluding domains and entities. If one needs to send only few data to ES, then she/he needs to add all other domains to the `exclude_domains` array.

**Describe the solution you'd like**
Add `include_domains` and `include_entities` with a precedence that included entities override excluded domains (and that excluded entities override include domains?).

### PR DESCRIPTION
**Pull request by ooii**
Please check the `CONFIG_SCHEMA`. 
The logic behind what I've done is the following:
If `included_domains` is not empty, then only the domains in it are included, all the others are excluded. The same logic is applied to entities (given that its domain is included).

Closes #49

### CODE REVIEW

**Comment by legrego** [comment]
@ooii thanks so much for putting this together! I'll try to take a closer look at this tomorrow and give you any feedback I have.

**Comment by legrego** [comment]
Sorry I haven't looked at this yet. I haven't forgotten!

File: custom_components/elastic.py [hunk 1]
```diff
@@ -273,8 +287,10 @@ def do_publish(self):
         if not self._only_publish_changed:
             all_states = self._hass.states.async_all()
             for state in all_states:
-                if (state.domain in self._excluded_domains
-                        or state.entity_id in self._excluded_entities):
+                if (not self._included_domains and state.domain in self._excluded_domains)
```
Comment (legrego) [inline custom_components/elastic.py]: This is complex enough that we should probably pull it out into its own function for readability

File: custom_components/elastic.py [hunk 2]
```diff
@@ -238,11 +250,13 @@ def enqueue_state(self, entry):
...
-        if domain in self._excluded_domains:
+        if (not self._included_domains and domain in self._excluded_domains) 
+            or (self._included_domains and domain not in self._included_domains):
             _LOGGER.debug("Skipping %s: it belongs to an excluded domain", entity_id)
             return
 
-        if entity_id in self._excluded_entities:
+        if (not self._included_entities and entity_id in self._excluded_entities) 
+            or (self._included_entities and entity_id not in self._included_entities):
             _LOGGER.debug("Skipping %s: this entity is explicitly excluded", entity_id)
```
Comment (legrego) [inline custom_components/elastic.py]: These log messages aren't technically correct in all cases. We'll want to either rephrase these messages, or have dedicated messages for the include and exclude scenarios.

File: custom_components/elastic.py (hunk 1 above)
Comment (ooii) [reply custom_components/elastic.py]: I rephrased debug messages about skipping entity/domain. The output now only states if a state is skipped or to be published.

File: custom_components/elastic.py (hunk 2 above)
Comment (ooii) [reply custom_components/elastic.py]: I took out the function and call it in `enqueue_state` and `do_publish`.

**Comment by ooii** [comment]
Sorry for the diff.

**Comment by legrego** [comment]
@ooii let me know if you're still interested in working on this. No hard feelings if not, I can finish this up if needed!

**Comment by ooii** [comment]
Hello,

Sorry @legrego for the response delay. It was the rush before leaving for vacation where I'm now. But I'm still interested to participating to that.

> One thing I'd still like to see is making the include.* and exclude.* configuration entries mutually exclusive. We shouldn't allow users to specify both include and exclude configurations when we will only respect one of them.

Actually, I think a user could include/exclude a domain and its entities without any issue. Example:
1. Include `binary_sensor` domain: that will include all the binary sensors. 
    1. If the user includes explicitly a `binary_sensor`, it won't change anything. 
    2. If the user excludes explicitly a `binary_sensor`, then this latter is excluded while all others are included.
2. Exclude `binary_sensor` domain: that will exclude all the binary sensors. 
    1. If the user excludes explicitly a `binary_sensor`, it won't change anything. 
    2. If the user includes explicitly a `binary_sensor`, then this latter is the only `binary_sensor` which is included.

**Note** that `2.i` could be obtained if the user only includes the specific `binary_sensor`, while 1.i should not be obtained if the user only excludes the specific `binary_sensor` (I'm not sure that excluding a specific entity should include all other entities of the domain).

> We shouldn't allow users to specify both include and exclude configurations when we will only respect one of them

If you mean that including and excluding the same domain/entity should not be accepted, then I agree.

**Comment by legrego** [comment]
> Sorry @legrego for the response delay. It was the rush before leaving for vacation where I'm now. But I'm still interested to participating to that.

No worries at all, and please don't feel rushed!

> Actually, I think a user could include/exclude a domain and its entities without any issue

Thanks for explaining your viewpoint here with examples, this is very helpful! I agree that it's useful to allow both include and exclude for the scenarios you listed above. I think we would want to prevent both `exclude.domains` and `include.domains` from being specified, and likewise, prevent both `exclude.entities` and `include.entities` from being specified.

Once `include.domains` or `include.entities` is specified, it declares a whitelist, so specifying the corresponding `exclude` configuration is a no-op, and might lead to confusion.

So in summary:

| exclude domain | exclude entity | include domain | include entity | Allowed? |
|---|---|---|---|---|
| - | - | - | - | Yes |
| Set | - | - | - | Yes |
| - | Set | - | - | Yes |

**Comment by ooii** [comment]
I completely agree with that. 
I'd suggest also to have the ability to use regex in include/exclude of entities.

**Comment by legrego** [comment]
> I'd suggest also to have the ability to use regex in include/exclude of entities.

I'm not opposed to this, as I think it'd be a pretty powerful feature. My only request is that we make it possible to easily specify entities using normal string notation (e.g., `switch.living_room_lights`) without having to escape regex characters (e.g., `switch\.living_room_lights`). Were you thinking of having a separate configuration entry to specify regex vs strings, or did you have something else in mind?

If it's easier, we can always add regex support in a followup PR too, if the string based approach isn't powerful enough.

**Comment by ooii** [comment]
>My only request is that we make it possible to easily specify entities using normal string notation (e.g., switch.living_room_lights) without having to escape regex characters (e.g., switch\.living_room_lights). 

Totally agreed with that.

>Were you thinking of having a separate configuration entry to specify regex vs strings, or did you have something else in mind?

Yes, I'm having this in mind this. We should replace (for `excluded_domains`)
```
if (not self._included_domains and state.domain in self._excluded_domains):
```
by
```
regex_excluded_domains = "(" + ")|(".join(self._excluded_domains) + ")"
if (not self._included_domains and re.search(regex_excluded_domains, state.domain)):
```

**Comment by legrego** [comment]
@ooii that sounds like a plan to me!

**Comment by ooii** [comment]
@legrego are you on it or do you want me to send a new PR?

**Comment by legrego** [comment]
@ooii I'll leave that up to you. I haven't done any work on this yet, so if you're still interested in working on this, feel free! Just let me know your plans 😄

**Comment by ooii** [comment]
That's fine for me. But I'd like to let you manage the part about include.* and exclude.* configuration entries mutually exclusive. Fine?

**Comment by legrego** [comment]
good with me!

**Comment by legrego** [comment]
I'm going to close this since it is over a year behind `master`. Please rebase and reopen if this is something you're still interested in working on. Thanks!
